test/unit/vm/cache/cache_test.dart

```diff
@@ -0,0 +1,301 @@
+        await cache.dispose();
+        expect(
+            () => cache.get(cachedId, () => notCachedValue), throwsStateError);
+      });
+    });
+
+    group('put', () {
+      test('should update cache', () async {
+        await cache.put(putId, putValue);
+        final getValue = await cache.get(putId, () => null);
+        expect(getValue, same(putValue));
+      });
+
+      test('should call onPut', () async {
+        final ChildCache childCache = new ChildCache();
+
+        await childCache.put(putId, putValue);
+        expect(childCache.onPutId, putId);
+        expect(childCache.onPutValue, putValue);
+      });
+
+      test('should put syncronously', () {
+        final getA = cache.getAsync(notCachedId, () async {
+          await new Future.delayed(new Duration(milliseconds: 100));
+          return notCachedValue;
+        });
+
+        final put = cache.put(notCachedId, putValue);
+        final getB = cache.get(notCachedId, () => new Object());
+
+        getA.then(expectAsync1(
+            (Object value) => expect(value, same(notCachedValue))));
+        getB.then(
+            expectAsync1((Object value) => expect(value, same(putValue))));
+        put.then(expectAsync1((Object value) => expect(value, same(putValue))));
+      });
+
+      test('should throw when disposed', () async {
+        await cache.dispose();
+        expect(() => cache.put(putId, null), throwsStateError);
+      });
+    });
+
+    group('remove', () {
+      test('should dispatch one didUpdate event when identifier is removed',
+          () async {
+        cache.didUpdate.listen(expectAsync1((CacheContext context) {
+          expect(context.id, cachedId);
+          expect(context.value, isNull);
+        }, count: 1));
+        await cache.remove(cachedId);
+        await cache.remove(cachedId);
+      });
+
+      test(
+          'should dispatch one didUpdate event when identifier is removed '
+          'syncronously', () {
+        cache.didUpdate.listen(expectAsync1((CacheContext context) {
+          expect(context.id, cachedId);
+          expect(context.value, isNull);
+        }, count: 1));
+        cache.remove(cachedId);
+        cache.remove(cachedId);
+      });
+
+      test('should dispatch one didRemove event when identifier is removed',
+          () async {
+        cache.didRemove.listen(expectAsync1((CacheContext context) {
+          expect(context.id, cachedId);
+          expect(context.value, cachedValue);
+        }, count: 1));
+        await cache.remove(cachedId);
+        await cache.remove(cachedId);
+      });
+
+      test(
+          'should dispatch one didRemove event when identifier is removed '
+          'syncronously', () {
+        cache.didRemove.listen(expectAsync1((CacheContext context) {
+          expect(context.id, cachedId);
+          expect(context.value, cachedValue);
+        }, count: 1));
+        cache.remove(cachedId);
+        cache.remove(cachedId);
```

Comment:
Would a future have to be awaited here? I don't know if this would fail after the test is already finished.